pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go

```diff
@@ -64,6 +78,115 @@ func (admitter *VMIUpdateAdmitter) Admit(ar *v1beta1.AdmissionReview) *v1beta1.A
 	return &reviewResponse
 }
 
+// admitHotplug compares the old and new volumes and disks, and ensures that they match and are valid.
+func admitHotplug(newVolumes []v1.Volume, newDisks []v1.Disk, volumeStatuses []v1.VolumeStatus) *v1beta1.AdmissionResponse {
+	if len(newVolumes) != len(newDisks) {
+		return webhookutils.ToAdmissionResponse([]metav1.StatusCause{
+			{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: "number of disks does not equal the number of volumes",
+			},
+		})
+	}
+	hotplugVolumeMap := getHotplugVolumes(newVolumes, volumeStatuses)
+	permanentVolumeMap := getPermanentVolumes(newVolumes, hotplugVolumeMap)
+
+	// Ensure we didn't remove non hot plugged disks and volumes.
+	permanentCount := 0
+	for _, volume := range newVolumes {
+		if _, ok := permanentVolumeMap[volume.Name]; ok {
+			permanentCount++
+		}
+	}
+	if len(permanentVolumeMap) > permanentCount || len(permanentVolumeMap) == 0 {
+		// Removed one of the permanent volumes, reject admission.
+		return webhookutils.ToAdmissionResponse([]metav1.StatusCause{
+			{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: "cannot remove permanent volume",
+			},
+		})
+	}
+	// Ensure the permanent volumes and disks are the first in the list.
```

Comment:
is this ordering specific to permanent volumes, or do we need to ensure that hotplugged volumes don't change order as well?

It seems like our concern here is around the device naming, like vdb, vdc, etc...  Maybe that's something we should keep track of on the VolumeStatus field for each volume so we don't have to worry about ordering after the initial vmi CREATE

Reply:
This is an extremely good point. I am thinking the following scenario, lets say we started with 2 disks, the root disk, and cloudinit. So we have vda and vdb. Then we hotplug sda/sdb/sbc (scsi), then we unplug sda. We don't want to change sdb and sdc, and if we keep track of that in the volume status, we can use it when calculating the target when plugging in the volume into the VM.